File: loopx/control_plane/coordination/coordination_state_contract.py
"""Python validation facade for the generated coordination-state contract."""

from __future__ import annotations

from typing import Mapping

from .coordination_state_contract_generated import COORDINATION_STATE_CONTRACT


COORDINATION_STATE_CONTRACT_SCHEMA_VERSION = "loopx_coordination_state_contract_v0"


class CoordinationStateContractError(ValueError):
    """A record or bundled contract violates the coordination-state schema."""


if COORDINATION_STATE_CONTRACT.get("schema_version") != (
    COORDINATION_STATE_CONTRACT_SCHEMA_VERSION
):
    raise CoordinationStateContractError("coordination state contract schema mismatch")


_TODO_READ_RECORD = COORDINATION_STATE_CONTRACT.get("todo_read_record")
if not isinstance(_TODO_READ_RECORD, Mapping):
    raise CoordinationStateContractError("Todo record contract must be an object")

TODO_CANONICAL_READ_RECORD_SCHEMA_VERSION = _TODO_READ_RECORD.get("schema_version")
TODO_ITEM_SCHEMA_VERSION = _TODO_READ_RECORD.get("item_schema_version")
if (
    not isinstance(TODO_CANONICAL_READ_RECORD_SCHEMA_VERSION, str)
    or not TODO_CANONICAL_READ_RECORD_SCHEMA_VERSION
    or not isinstance(TODO_ITEM_SCHEMA_VERSION, str)
    or not TODO_ITEM_SCHEMA_VERSION
):
    raise CoordinationStateContractError("Todo record schema versions must be strings")
TODO_CANONICAL_READ_RECORD_FIELDS = tuple(_TODO_READ_RECORD["fields"])
TODO_CANONICAL_REQUIRED_READ_FIELDS = tuple(_TODO_READ_RECORD["required_fields"])

_DOMAIN = COORDINATION_STATE_CONTRACT["todo_domain_record"]
_PROJECTION = COORDINATION_STATE_CONTRACT["todo_projection_metadata"]
if (
    _DOMAIN.get("fields_from") != "todo_read_record"
    or _DOMAIN.get("exclude_fields_from") != "todo_projection_metadata"
):
    raise CoordinationStateContractError("Todo domain contract mismatch")
TODO_PROJECTION_METADATA_FIELDS = tuple(_PROJECTION["fields"])
TODO_DOMAIN_READ_RECORD_SCHEMA_VERSION: str = _DOMAIN["schema_version"]
TODO_DOMAIN_ITEM_SCHEMA_VERSION: str = _DOMAIN["item_schema_version"]
TODO_DOMAIN_RECORD_FIELDS = tuple(
    field for field in TODO_CANONICAL_READ_RECORD_FIELDS
    if field not in TODO_PROJECTION_METADATA_FIELDS
)
TODO_DOMAIN_REQUIRED_FIELDS = tuple(_DOMAIN["required_fields"])
if not set(TODO_DOMAIN_REQUIRED_FIELDS) <= set(TODO_DOMAIN_RECORD_FIELDS):
    raise CoordinationStateContractError("Todo domain required fields are not declared")


def canonical_record_fields(
    value: Mapping[str, object],
    *,
    fields: tuple[str, ...],
    required_fields: tuple[str, ...],
    label: str,
    reject_unknown: bool,
) -> dict[str, object]:
    """Copy one record through the shared field contract.

    ``reject_unknown`` is mandatory for provider-bound projections.  The false
    mode exists only for legacy status callers that may carry presentation-only
    fields outside the machine-owned coordination record.
    """

    unknown_required = sorted(set(required_fields).difference(fields))
    if unknown_required:
        raise CoordinationStateContractError(
            f"{label} required fields are absent from fields: "
            + ", ".join(unknown_required)
        )
    if reject_unknown:
        unexpected = sorted(set(value).difference(fields))
        if unexpected:
            raise CoordinationStateContractError(
                f"{label} has unversioned fields: {', '.join(unexpected)}"
            )
    record = {
        field: value[field]
        for field in fields
        if field in value and value[field] is not None
    }
    missing = [field for field in required_fields if field not in record]
    if missing:
        raise CoordinationStateContractError(
            f"{label} omits required fields: {', '.join(missing)}"
        )
    return record


__all__ = [
    "COORDINATION_STATE_CONTRACT",
    "COORDINATION_STATE_CONTRACT_SCHEMA_VERSION",
    "CoordinationStateContractError",
    "TODO_CANONICAL_READ_RECORD_FIELDS",
    "TODO_CANONICAL_READ_RECORD_SCHEMA_VERSION",
    "TODO_CANONICAL_REQUIRED_READ_FIELDS",
    "TODO_ITEM_SCHEMA_VERSION",
    "TODO_DOMAIN_READ_RECORD_SCHEMA_VERSION",
    "TODO_DOMAIN_ITEM_SCHEMA_VERSION",
    "TODO_DOMAIN_RECORD_FIELDS",
    "TODO_DOMAIN_REQUIRED_FIELDS",
    "TODO_PROJECTION_METADATA_FIELDS",
    "canonical_record_fields",
]

File: loopx/control_plane/coordination/coordination_state_contract.ts
import type { JsonObject } from "../effect_program.ts";
import {
  AuthorityStoreProtocolError,
  canonicalAuthorityObject,
} from "./authority_store_codec.ts";
import { COORDINATION_STATE_CONTRACT } from "./coordination_state_contract.generated.ts";

export const COORDINATION_STATE_CONTRACT_SCHEMA =
  "loopx_coordination_state_contract_v0";

if (COORDINATION_STATE_CONTRACT.schema_version !== COORDINATION_STATE_CONTRACT_SCHEMA) {
  throw new AuthorityStoreProtocolError("coordination state contract schema mismatch");
}

interface RecordContract {
  readonly fields: readonly string[];
  readonly required_fields: readonly string[];
}

export { COORDINATION_STATE_CONTRACT };
export const TODO_CANONICAL_READ_RECORD_SCHEMA =
  COORDINATION_STATE_CONTRACT.todo_read_record.schema_version;
export const TODO_ITEM_SCHEMA =
  COORDINATION_STATE_CONTRACT.todo_read_record.item_schema_version;
export const TODO_CANONICAL_READ_RECORD_FIELDS =
  COORDINATION_STATE_CONTRACT.todo_read_record.fields;
export const TODO_CANONICAL_REQUIRED_READ_FIELDS =
  COORDINATION_STATE_CONTRACT.todo_read_record.required_fields;

// The legacy manifest remains an immutable compatibility contract. Native
// provider records have their own version and never require Markdown location.
export const TODO_DOMAIN_READ_RECORD_SCHEMA =
  COORDINATION_STATE_CONTRACT.todo_domain_record.schema_version;
export const TODO_DOMAIN_ITEM_SCHEMA =
  COORDINATION_STATE_CONTRACT.todo_domain_record.item_schema_version;
export const TODO_DOMAIN_RECORD_CONTRACT: RecordContract = Object.freeze({
  fields: Object.freeze(TODO_CANONICAL_READ_RECORD_FIELDS.filter((field) =>
    !(COORDINATION_STATE_CONTRACT.todo_projection_metadata.fields as readonly string[]).includes(field))),
  required_fields: COORDINATION_STATE_CONTRACT.todo_domain_record.required_fields,
});

export interface TodoDomainRecord extends JsonObject {
  schema_version: string;
  todo_id: string;
  role: "user" | "agent";
  status: "open" | "done" | "blocked" | "deferred";
  done: boolean;
  text: string;
  archive_state: "active" | "archive";
}

export interface TodoProjectionMetadata {
  readonly source_section: string;
  readonly index?: number;
}

export function canonicalTodoDomainRecord(value: unknown, label = "Todo domain record"): TodoDomainRecord {
  const record = canonicalCoordinationRecord(value, TODO_DOMAIN_RECORD_CONTRACT, label);
  const terminal = record.status === "done" || record.status === "deferred";
  if (record.schema_version !== TODO_DOMAIN_ITEM_SCHEMA ||
      typeof record.todo_id !== "string" || !record.todo_id ||
      (record.role !== "user" && record.role !== "agent") ||
      typeof record.status !== "string" ||
      !["open", "done", "blocked", "deferred"].includes(String(record.status)) ||
      typeof record.done !== "boolean" || record.done !== terminal ||
      typeof record.text !== "string" ||
      (record.archive_state !== "active" && record.archive_state !== "archive")) {
    throw new AuthorityStoreProtocolError(`${label} has invalid required semantics`);
  }
  return record as TodoDomainRecord;
}

export function canonicalCoordinationRecord(
  value: unknown,
  contract: RecordContract,
  label: string,
): JsonObject {
  const record = canonicalAuthorityObject(value, label);
  const allowed = new Set(contract.fields);
  const unknownRequired = contract.required_fields.filter(
    (field) => !allowed.has(field),
  );
  if (unknownRequired.length > 0) {
    throw new AuthorityStoreProtocolError(
      `${label} required fields are absent from fields: ${unknownRequired.join(", ")}`,
    );
  }
  const unexpected = Object.keys(record)
    .filter((field) => !allowed.has(field))
    .sort((left, right) => left.localeCompare(right));
  if (unexpected.length > 0) {
    throw new AuthorityStoreProtocolError(
      `${label} has unversioned fields: ${unexpected.join(", ")}`,
    );
  }
  const missing = contract.required_fields.filter((field) => !(field in record));
  if (missing.length > 0) {
    throw new AuthorityStoreProtocolError(
      `${label} omits required fields: ${missing.join(", ")}`,
    );
  }
  return canonicalAuthorityObject(Object.fromEntries(
    contract.fields.flatMap((field) => field in record ? [[field, record[field]]] : []),
  ), label);
}

export function canonicalCoordinationTodoRecord(
  value: unknown,
  label = "coordination Todo read record",
): JsonObject {
  const record = canonicalCoordinationRecord(
    value,
    COORDINATION_STATE_CONTRACT.todo_read_record,
    label,
  );
  if (record.schema_version !== TODO_ITEM_SCHEMA ||
      (record.role !== "user" && record.role !== "agent") ||
      typeof record.status !== "string" || record.status.length === 0 ||
      typeof record.done !== "boolean" || typeof record.text !== "string" ||
      typeof record.archive_state !== "string" || record.archive_state.length === 0 ||
      typeof record.source_section !== "string" || record.source_section.length === 0) {
    throw new AuthorityStoreProtocolError(`${label} has invalid required semantics`);
  }
  return record;
}
